lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
+	buffer.Write(nibblesToKeyLE(b.key))
+	buffer.Write(common.Uint16ToBytes(b.childrenBitmap()))
 
 	if b.value != nil {
 		bytes, err := scale.Marshal(b.value)
 		if err != nil {
-			return nil, err
-		}
-		h.tmp.Write(bytes)
-	}
-
-	if h.parallel {
-		wg, _ := errgroup.WithContext(context.Background())
-		resBuff := make([][]byte, 16)
-		for i := 0; i < 16; i++ {
-			func(i int) {
-				wg.Go(func() error {
-					child := b.children[i]
-					if child == nil {
-						return nil
-					}
-
-					var err error
-					resBuff[i], err = encodeAndHash(child)
-					if err != nil {
-						return err
-					}
-					return nil
-				})
-			}(i)
-		}
-		if err := wg.Wait(); err != nil {
-			return nil, err
+			return err
 		}
+		buffer.Write(bytes)
+	}
+
+	if parallel {
+		return encodeChildsInParallel(b.children, buffer)
+	}
+	return encodeChildsSequentially(b.children, buffer)
+}
 
-		for _, v := range resBuff {
-			if v != nil {
-				h.tmp.Write(v)
+func encodeChildsInParallel(children [16]node, buffer *bytes.Buffer) (err error) {
+	type result struct {
+		index  int
+		buffer *bytes.Buffer
+		err    error
+	}
+
+	resultsCh := make(chan result)
+
+	for i, child := range children {
+		go func(index int, child node) {
+			buffer := encodingBufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
+			// buffer is put back in the pool after processing its
+			// data in the select block below.
+
+			err := encodeChild(child, buffer)
+
+			resultsCh <- result{
+				index:  index,
+				buffer: buffer,
+				err:    err,
 			}
+		}(i, child)
+	}
+
+	currentIndex := 0
+	resultBuffers := make([]*bytes.Buffer, len(children))
+	for range children {
+		result := <-resultsCh
+		if result.err != nil && err == nil { // only set the first error we get
+			err = result.err
 		}
-	} else {
-		for i := 0; i < 16; i++ {
-			if child := b.children[i]; child != nil {
-				scEncChild, err := encodeAndHash(child)
-				if err != nil {
-					return nil, err
-				}
-				h.tmp.Write(scEncChild)
+
+		resultBuffers[result.index] = result.buffer
+
+		// write as many completed buffers to the result buffer.
+		for currentIndex < len(children) &&
+			resultBuffers[currentIndex] != nil {
+			// note buffer.Write copies the byte slice given as argument
+			_, writeErr := buffer.Write(resultBuffers[currentIndex].Bytes())
+			if writeErr != nil && err == nil {
+				err = writeErr
 			}
+
+			encodingBufferPool.Put(resultBuffers[currentIndex])
+			resultBuffers[currentIndex] = nil
+
+			currentIndex++
 		}
 	}
 
-	return h.tmp.Bytes(), nil
+	for _, buffer := range resultBuffers {
+		if buffer == nil { // already emptied and put back in pool
+			continue
+		}
+		encodingBufferPool.Put(buffer)
+	}
+
+	return err
 }
 
-// encodeLeaf encodes a leaf with the encoding specified at the top of this package
-func (h *hasher) encodeLeaf(l *leaf) ([]byte, error) {
-	if !l.dirty && l.encoding != nil {
-		return l.encoding, nil
+func encodeChildsSequentially(children [16]node, buffer *bytes.Buffer) (err error) {
```

Comment:
```suggestion
func encodeChildrenSequentially(children [16]node, buffer *bytes.Buffer) (err error) {
```

Reply:
done in 99f273dc1c9e1b69d30de3995b37dd349b4769fd